commons-vfs2-sandbox/src/main/java/org/apache/commons/vfs2/provider/smb2/Smb2ClientWrapper.java

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.commons.vfs2.provider.smb2;
+
+import com.hierynomus.msdtyp.AccessMask;
+import com.hierynomus.msfscc.FileAttributes;
+import com.hierynomus.msfscc.fileinformation.FileAllInformation;
+import com.hierynomus.msfscc.fileinformation.FileIdBothDirectoryInformation;
+import com.hierynomus.mssmb2.SMB2CreateDisposition;
+import com.hierynomus.mssmb2.SMB2CreateOptions;
+import com.hierynomus.mssmb2.SMB2ShareAccess;
+import com.hierynomus.mssmb2.SMBApiException;
+import com.hierynomus.smbj.SMBClient;
+import com.hierynomus.smbj.SmbConfig;
+import com.hierynomus.smbj.auth.AuthenticationContext;
+import com.hierynomus.smbj.connection.Connection;
+import com.hierynomus.smbj.session.Session;
+import com.hierynomus.smbj.share.DiskEntry;
+import com.hierynomus.smbj.share.DiskShare;
+import org.apache.commons.vfs2.FileSystemException;
+import org.apache.commons.vfs2.FileSystemOptions;
+import org.apache.commons.vfs2.provider.GenericFileName;
+
+import java.util.ArrayList;
+import java.util.EnumSet;
+import java.util.List;
+
+/**
+ * A wrapper to the SMBClient for bundling the related client & connection instances.
+ */
+public class Smb2ClientWrapper extends SMBClient {
+
+    private static final SmbConfig CONFIG = SmbConfig.builder()
+            .withDfsEnabled(true)
+            .withMultiProtocolNegotiate(true)
+            .build();
+
+    protected final FileSystemOptions fileSystemOptions;
+    private final GenericFileName root;
+    private SMBClient smbClient;
+    private Connection connection;
+    private Session session;
+    private DiskShare diskShare;
+
+    protected Smb2ClientWrapper(final GenericFileName root, final FileSystemOptions fileSystemOptions)
+            throws FileSystemException {
+
+        this.root = root;
+        this.fileSystemOptions = fileSystemOptions;
+        smbClient = new SMBClient(CONFIG);
+        setupClient();
+    }
+
+    private void setupClient() throws FileSystemException {
+
+        final GenericFileName rootName = getRoot();
+
+        //the relevant data to authenticate a connection
+        String userName;
+        if (rootName.getUserName().equals("") || rootName.getUserName() == null) {
+            userName = "";
+        } else {
+            if (rootName.getUserName().contains(";")) {
```

Comment:
Shall we define a const for ; and reuse. Also shall we assign rootName.getUserName() to a variable and re use.

Reply:
Done